runtime/lua/vim/lsp/codelens.lua

```diff
@@ -8,14 +8,28 @@ local M = {}
 --- to throttle refreshes to at most one at a time
 local active_refreshes = {}
 
---- bufnr -> client_id -> lenses
+---@type table<integer, table<integer, lsp.CodeLens[]>>
+--- valid bufnr → client_id → lenses.
+--- Invariant: If a lens is displayed, then lens_cache_by_buf[bufnr][client_id] is not nil.
 local lens_cache_by_buf = setmetatable({}, {
   __index = function(t, b)
     local key = b > 0 and b or api.nvim_get_current_buf()
     return rawget(t, key)
   end,
 })
 
+---@param bufnr integer
+---@param client_id integer
+local function ensure_init(bufnr, client_id)
```

Comment:
woud this function be needed if `lens_cache_by_buf` were defined as a [vim.defaulttable()](https://neovim.io/doc/user/lua.html#vim.defaulttable()) ?

Reply:
No. For `clear(nil, nil)` to be correct, something should keep track of bufs/clients that have lenses. But there are two places where lens (virtual text) are added without touching `lens_cache_by_buf`.
* `display()`: User may add lenses arbitrarily. For this case, we can simply add the lens to the cache (instead of `ensure_init` thing).
* `'codeLens/resolve'` handler: Since it's asynchronously executed, when the handler is executed, `lens_cache_by_buf` might have been cleared already. In this case, we need to check whether the lens should be added to avoid duplicates in `get()` result.

Reply:
is there a 1-2 line comment that can explain that here, to help others in the future:

    // Note: lens_cache_by_buf cannot be vim.defaulttable() because...

Reply:
> something should keep track of bufs/clients that have lenses. 

It seems that this isn't necessary. nvim_list_buf() and vim.tbl_keys(namespaces) should be fine.

Reply:
I have simplified this stuff